File: ReplaceWords648.java
/* Time Complexity : O(N*M)for inserting into trie + O(N) for splitting words into tokens + 0(KM) for searching k words of length M in trie.
 Space Complexity :  O(K*M) for storing results + split sentence to tokens + O(N*M) to build a trie from dictionary.
 where N is dictionary size and M is length of each string.
 Did this code successfully run on Leetcode : Yes
 Any problem you faced while coding this : No


/*
1. Build a trie using words given in dictionary
2. split the words in sentence and search if the part of it exists as a word (not prefix) in trie.
3. Add it to result if exists, otherwise add same word.
*/
class Solution {
    class TrieNode{
        TrieNode children[];
        boolean is_end;
        TrieNode(){
            children = new TrieNode[26];
            is_end = false;
        }
    }
    TrieNode root = new TrieNode();
    void insert(String word){
        TrieNode curr;
        curr = root;
        for(int i =0 ; i<word.length(); i++){
            int index = word.charAt(i) - 'a';
            if(curr.children[index] == null){
                curr.children[index] = new TrieNode();
            }
            curr = curr.children[index];
        }
        curr.is_end = true;
    }
    public String replaceWords(List<String> dictionary, String sentence) {
        for(String word : dictionary){
            insert(word);
        }
        String words[] = sentence.split(" ");
        StringBuilder result = new StringBuilder();
        for(int i = 0 ; i<words.length ; i++){
            if( i != 0){
                result.append(" ");
            }
          StringBuilder replacemnt  = new StringBuilder();
         String currentString = words[i];
         TrieNode curr = root;
            for( int j = 0 ; j<currentString.length() ; j++){
                int index = currentString.charAt(j) - 'a';
                if(curr.children[index] == null || curr.is_end) break;
                replacemnt.append(currentString.charAt(j));
                curr = curr.children[index];
            }
            if(curr.is_end){
                result = result.append(replacemnt);
            }else{
                result = result.append(currentString);
            }
        }
    
        return result.toString();
        
    }
}

File: replace-words_648.py
#Using trie
#Time Complexity O(M * l)  for dictionary words of average l length, and O(N * k) for fpr number of words of k length
#Space Complexity O(M * N)
#It successfully runs on leetcode 
class TriNode:
    def __init__(self):
        self.children = [None for i in range(26)]
        self.isend = False
        
class Solution:
    def __init__(self):
        self.root = TriNode()
    
    def insert(self, word):
        cur = self.root
        for c in word:
            if cur.children[ord(c)-ord("a")] == None:
                cur.children[ord(c)-ord("a")] = TriNode()
            cur = cur.children[ord(c)-ord("a")]
        cur.isend = True
                                        
                                
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        for i in dictionary:
            self.insert(i)
        st = sentence.split(" ")
        res = ""
        firstword = True
        for word in st:
            cur = self.root
            rs= ""
            for c in word:
                if cur.children[ord(c)-ord("a")] == None or  cur.isend==True:
                    break
                rs+=c
                cur = cur.children[ord(c)-ord("a")]
                
            if firstword: 
                res = res + rs if cur.isend else res+word
                firstword = False
            else:
                res = res + " " + rs if cur.isend else res + " " + word
        return res                            

#Approach: Using Hashset                                            
#Time Complexity O(M * l)  for dictionary words of average l length, and O(N * k^2) for fpr number of words of k length
#Space Complexity O(M)
#It successfully runs on leetcode 
        
class Solution:      
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        s = set()
        for i in dictionary:
            s.add(i) 
        st = sentence.split(" ")
        res = ""
        firstword = True
        for word in st:           
            prevlen = len(res)
            for c in range(len(word)):
                sub = word[:c]
                if sub in s:
                    res= res + sub  if firstword else res + " " + sub
                    break
            if prevlen==len(res):
                res= res + word if firstword else res + " " + word
            firstword = False
        return res                            
                            